### PR TITLE
chore: roll to tot

### PR DESCRIPTION
**Pull request by kumaraditya303**
Will be followed by other PR

### CODE REVIEW

File: binding_call.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ type BindingSource struct {
 type ExposedFunction = func(args ...interface{}) interface{}
 
 // BindingCallFunction represents the func signature of an exposed binding call func
-type BindingCallFunction = func(source *BindingSource, args ...interface{}) interface{}
+type BindingCallFunction func(source *BindingSource, args ...interface{}) interface{}
```
Comment (mxschmitt) [inline binding_call.go:22]: whats the reason that this had to be changed? Just curious.

File: transport.go [hunk 2]
```diff
@@ -7,92 +7,17 @@ import (
 	"fmt"
 	"io"
 	"log"
-	"net/http"
 	"os"
 	"sync"
-	"sync/atomic"
-	"time"
 
-	"github.com/gorilla/websocket"
```
Comment (mxschmitt) [inline transport.go:16]: please remove this dependency from the go.mod file?

File: tests/route_test.go [hunk 3]
```diff
@@ -173,6 +173,7 @@ func TestRouteFulfillPath(t *testing.T) {
 
 func TestRequestFinished(t *testing.T) {
 	BeforeEach(t)
+	t.Skip("TODO: fix this test")
```
Comment (mxschmitt) [inline tests/route_test.go:176]: how is this failing?

File: objectFactory.go [hunk 4]
```diff
@@ -44,6 +44,12 @@ func createObjectFactory(parent *channelOwner, objectType string, guid string, i
 		return nil
 	case "Electron":
 		return nil
+	case "FetchRequest":
+		c := &channelOwner{}
```
Comment (mxschmitt) [inline objectFactory.go:48]: maybe `return nil` for now?

File: run.go [hunk 5]
```diff
@@ -239,16 +246,9 @@ func Run(options ...*RunOptions) (*Playwright, error) {
...
-		}
-	}()
-	obj, err := connection.CallOnObjectWithKnownName("Playwright")
-	if err != nil {
-		return nil, fmt.Errorf("could not call object: %w", err)
-	}
-	return obj.(*Playwright), nil
+	connection.Start()
+	playwright := <-connection.playwright
+	return playwright, nil
```
Comment (mxschmitt) [inline run.go:251]:
```suggestion
	return <-connection.playwright, nil
```

File: jsonPipe.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+type jsonPipe struct {
+	channelOwner
+}
+
+func (j *jsonPipe) Send(message map[string]interface{}) error {
+	if _, err := j.channel.Send("send", map[string]interface{}{
+		"message": message,
+	}); err != nil {
```
Comment (mxschmitt) [inline jsonPipe.go:14]: nit: no if needed here. Just return err.

File: jsonPipe.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+		return err
+	}
+	return nil
+}
+
+func (j *jsonPipe) Close() error {
+	if _, err := j.channel.Send("close"); err != nil {
```
Comment (mxschmitt) [inline jsonPipe.go:21]: dito

File: jsonPipe.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	}
+	return nil
+}
+func newJsonPipe(parent *channelOwner, objectType string, guid string, initializer map[string]interface{}) *jsonPipe {
+	j := &jsonPipe{}
+	j.createChannelOwner(j, parent, objectType, guid, initializer)
+	j.channel.On("message", func(ev map[string]interface{}) {
+		m, _ := json.Marshal(ev["message"])
+		var msg message
+		_ = json.Unmarshal(m, &msg)
```
Comment (mxschmitt) [inline jsonPipe.go:32]: nit: please log the error to the stdout via log.Fatalf if there is an error happening.

File: network.go [hunk 9]
```diff
@@ -10,3 +10,45 @@ func parseHeaders(headers []interface{}) map[string]string {
...
+type rawHeaders struct {
+	headersArray map[string]string
+	headersMap   map[string][]string
+}
+
+func (r *rawHeaders) Get(name string) string {
+	values := r.GetAll(name)
+	sep := ", "
+	if strings.ToLower(name) == "set-cookie" {
+		sep = "; "
```
Comment (mxschmitt) [inline network.go:23]:
```suggestion
		sep = "\n"
```
at least we have that upstream.

File: network.go [hunk 10]
```diff
@@ -10,3 +10,45 @@ func parseHeaders(headers []interface{}) map[string]string {
...
+	}
+	return strings.Join(values, sep)
+}
+func (r *rawHeaders) GetAll(name string) []string {
+	return r.headersMap[strings.ToLower(name)]
+}
+func (r *rawHeaders) Headers() map[string]string {
+	out := make(map[string]string)
+	for key, value := range r.headersArray {
```
Comment (mxschmitt) [inline network.go:32]: this should loop over the keys of headersMap and use the r.Get method instead.

File: network.go [hunk 11]
```diff
@@ -10,3 +10,45 @@ func parseHeaders(headers []interface{}) map[string]string {
...
+	r := &rawHeaders{}
+	r.headersArray = parseHeaders(headers)
+	r.headersMap = make(map[string][]string)
+	for _, header := range headers {
+		entry := header.(map[string]interface{})
+		name := strings.ToLower(entry["name"].(string))
+		if _, ok := r.headersMap[name]; !ok {
+			r.headersMap[name] = make([]string, 0)
+			r.headersMap[name] = append(r.headersMap[name], entry["value"].(string))
+		}
```
Comment (mxschmitt) [inline network.go:51]:
```suggestion
		if _, ok := r.headersMap[name]; !ok {
			r.headersMap[name] = make([]string, 0)
		}
		r.headersMap[name] = append(r.headersMap[name], entry["value"].(string))
```
shouldn't it be like that?

File: binding_call.go (hunk 1 above)
Comment (kumaraditya303) [reply binding_call.go:22]: The go compiler wasn't able to compile this :/

File: tests/route_test.go (hunk 3 above)
Comment (kumaraditya303) [reply tests/route_test.go:176]: Response.Finished is broken and will fix in follow up

File: objectFactory.go (hunk 4 above)
Comment (kumaraditya303) [reply objectFactory.go:48]: this is required, in a follow up this will be fixed

File: binding_call.go (hunk 1 above)
Comment (mxschmitt) [reply binding_call.go:22]: oh interesting. LGTM.

File: network.go (hunk 10 above)
Comment (kumaraditya303) [reply network.go:32]: In a follow up will fix all network related

File: transport.go (hunk 2 above)
Comment (kumaraditya303) [reply transport.go:16]: It is used in testing ws

**Review by mxschmitt — APPROVED** [review]

**Pull request merged**